### ISSUE TITLE
Concurrency and Asynchrony: Futures

### ISSUE THREAD

**Issue by UsatiyNyan** [UsatiyNyan/serious-execution-library#4]
# Futures

sources:
- STL implementation of std::future, e.g. [Microsoft's](https://github.com/microsoft/STL/blob/main/stl/inc/future)
- [Futures for C++11 at Facebook (Folly)](https://engineering.fb.com/2015/06/19/developer-tools/futures-for-c-11-at-facebook/)
- [Lazy evaluation](https://wiki.haskell.org/Lazy_evaluation)

We are going to start from a "naive" implementation of the Future in C++ and look at how STL treats asynchrony in general.
The final version of the Future will be akin to something found in Haskell or other functional programming languages.
This is the foundation from which concepts like Sender/Receiver are born.

## std::future

A typical implementation of `<future>` contains the following:
- [refcounted](https://github.com/microsoft/STL/blob/main/stl/inc/future#L250C1-L252C29) [state](https://github.com/microsoft/STL/blob/main/stl/inc/future#L424-L425) protected under [mutex](https://github.com/microsoft/STL/blob/main/stl/inc/future#L426)
  - result availability is notified and listened through [condvar](https://github.com/microsoft/STL/blob/main/stl/inc/future#L427)
- handles to state like `promise` and `future`

The design is pretty simple:
- [for the std::future](https://github.com/microsoft/STL/blob/main/stl/inc/future#L300-L327) - when performing `future::get()`, under mutex check readiness and wait on condvar until result is available - essentially locking the current OS-thread
- [for the std::promise](https://github.com/microsoft/STL/blob/main/stl/inc/future#L340-L385) - when performing `promise::set_value()` or `promise::set_exception()`, lock the mutex and `notify_all` readiness

There are several problems with this approach:
- due to the design decision, that `future` and `promise` could outlive each other, and aren't required to fullfill or acquire the result during their lifetimes - access to the state is not _ordered_ and needs to be synchronized
- the mechanism to support continuation is not present, the typical solution is outlined in [the same header](https://github.com/microsoft/STL/blob/main/stl/inc/future#L1489-L1493):
  - start a separate thread that waits for the result from `future`
  - perform the callback after `future` is done
  - which, according to https://github.com/UsatiyNyan/serious-execution-library/issues/2, is pretty wasteful

## Future without locks

In this section we would construct a Future similar to [Folly](https://engineering.fb.com/2015/06/19/developer-tools/futures-for-c-11-at-facebook/).

To solve locking we would need to change the design that allowed such a solution to exist.
Let's propose the following:
1. `future` and `promise` are always created simultaneously:
```cpp
auto [f, p] = make_contract<ValueT>();
``` 
2. disallow `future::get()` and instead set the continuation for the `future` via a callback (set_callback is non-blocking):
```cpp
// thread 1
std::move(f).set_callback([](auto value){ ... }); // non-blocking
// thread 2
std::move(p).set_value(value); // non-blocking
``` 
3. require both `future` and `promise` to perform `set_callback` and `set_value` exactly once during their lifetime.

That way access synchronization to shared state and lifetime management could be simplified down to a single atomic variable. Whilst this API is more restricting - it still allows the same expressiveness (if not more thanks to callbacks) as `std::future`, for example we could reimplement blocking `future::get`:
```cpp
template <typename T>
T get_blocking(future f) {
    std::atomic<bool> event{false};
    std::optional<T> result;
    std::move(f).set_callback([&event](auto value) {
        result.emplace(value);
        event.store(true);
        event.notify_one();
    });
    event.wait_for(true);
    return result.value();
}
```
And it does not require a mutex for synchronization, but still blocks current thread.
This design is easily extendable towards composable Futures (parallel and sequential combinators such as `map`, `all`, `any`, etc.):
```cpp
template <typename T, typename U>
future<U> map(future<T> a_future, function<U(T)> a_function) {
    auto [new_future, new_promise] = make_contract<U>();
    a_future.set_callback([a_function, new_promise](T value) {
        new_promise.set_value( a_function(value) );
    });
    return new_future;
}
```

## Future without locks: impl

As it was noted above - both `future` and `promise` have a handle on shared state called `context`, let it be just a pointer:
```cpp
template <typename T>
std::pair<future, promise> make_contract() {
    auto* ptr = new context<T>{};
    return { future{ ptr }, promise { ptr } }; 
}
```

The possible states for the `context` are:
- `created`
- `has_value`
- `has_callback`
And possible transitions are only:
- `created` -> `has_value`
- `created` -> `has_callback`

If `set_value` encounters `has_callback` or `set_callback` encounters `has_value` - we can just perform the callback immediately and cleanup `context`. And obviously, since future and promise can _race_ - `state` should be accessed atomically.

```cpp
enum state : std::uint32_t {
    created, has_value, has_callback
};

template <typename T>
struct context {
    optional<T> v_; // optional only to be able to initlize without calling default c-tors
    function<void(T)> cb_;
    atomic<state> s_;
    
public:
    void set_value(T value) {
        v_.emplace(value); // safe: set_callback will not touch this field until we set `has_value`

        auto expected = state::created; // expect that we are first
        if (s_.compare_exchange_strong(expected, state::has_value)) {
            // set_value is first, no need to touch anything
            return;
        }

        // set_value has lost the _race_ - needs to fulfill callback and cleanup
        assert(expected == state::has_callback);

        cb_(v_.value());
        delete this;
    }

    // set_callback is the exact mirror of set_value <- excercise for the reader :)
};
```

> NOTE: due to the _race_ between `set_value` and `set_callback` - continuation could run on either thread. 

## Future without allocations

Whilst the proposed solution is flexible and promotes functional design, it still requires that shared state is allocated, since both `future` and `promise` _race_ during their _performing of the contract_. But we could observe, that before the asynchronous work is even started - we could accomodate a stable address for the result of asynchronous work on the stack.

This could work only if the asynchronous work itself is _lazy_ and triggered after consumer is ready. Which dictates a strict order:
- consumer prepares callbacks
- asynchronous work starts
- producer fulfills

That way we eliminate the need for refcounting, because the only allowed state transitions are:
- `created` -> `has_callback` -> `has_value` -> `destroyed`

Since all callbacks should be composed before the asynchronous work starts, we can delay creation of `context` until the whole Future pipeline has been setup. And that's how basics of Sender/Receiver frameworks are usually introduced (including the one in STL - `<execution>`):
- Sender ([`SomeSignal`](https://github.com/UsatiyNyan/serious-execution-library/blob/main/include/sl/exec/model/concept.hpp) in this library) - is a source of asynchrony, used to compose the asynchronous execution pipeline, produces Operation State
- Receiver ([`slot`](https://github.com/UsatiyNyan/serious-execution-library/blob/main/include/sl/exec/model/slot.hpp) in this library) - a callback that implements part of the asynchronous execution pipeline, usually considered to be an implementation detail
- Operation State ([`Connection`](https://github.com/UsatiyNyan/serious-execution-library/blob/main/include/sl/exec/model/concept.hpp) in this library) - result of connecting Sender and Receiver, once `start`-ed - contains a stable address for result of the asynchornous work and it's lifetime should be longer than the work it starts.

```cpp
template <typename R, typename T>
concept Receiver = requires(R receiver, T value) {
    receiver.set_value(value);
};

template <typename OS>
concept OperationState = requires(OS op_state) {
    op_state.start();
};

template <typename S, typename T>
concept Sender = requires(S sender, dummy_receiver<T> receiver) {
    sender.connect(receiver) -> OperationState;
};

// sender example: on_thread

template <typename T, Receiver<T> R>
struct on_thread_op_state {
    function<T()> f;
    R receiver;
    
    void start() {
        std::thread{
            [this] {
                receiver.set_value( f() );
            }
        }.detach();
    }
};

template <typename T>
struct on_thread_sender {
    function<T()> f;

    OperationState auto connect(Receiver<T> auto receiver) { 
        return on_thread_op_state{ f, receiver };
    }
};

// receiver example: get_blocking

template <typename T>
struct get_blocking_receiver {
    atomic<bool> event{false};
    optional<T> result;
    
    void set_value(T value) {
        result.emplace(value);
        event.store(true);
        event.notify_one();
    }
};

template <typename T>
T get_blocking(Sender<T> auto sender) {
    auto op_state = sender.connect( get_blocking_receiver{} ); // stable address
    op_state.start();
    op_state.receiver.event.wait_for(true);
    return op_state.receiver.result.value();
}

// usage example:
on_thread_sender<int> sender{ []{
    // think really hard
    return 42;
} };
int result = get_blocking(sender);
```

Notice that for the whole pipeline no allocations were used, and atomic synchronization is only implemented when the thread has to wait for asynchronous work on a different thread.

Usually also a Scheduler is introduced ([`executor`](https://github.com/UsatiyNyan/serious-execution-library/blob/main/include/sl/exec/model/executor.hpp) in this library) - it is used to plan and execute a step of the `receiver.set_value`. Scheduler is not explicitly required to run work on a parallel thread - work could be scheduled inline or on a pseudo thread.

This design is highly composable, to add a step to the asynchronous pipeline one would add a Sender, that injects a Receiver, e.g. [`map`](https://github.com/UsatiyNyan/serious-execution-library/blob/main/include/sl/exec/algo/tf/seq/map.hpp):
```cpp
template <typename U, typename T, Receiver<U> R>
struct map_receiver {
    R receiver;
    function<U(T)> f;

    void set_value(T value) {
        auto result = f(value);
        // here we would schedule `set_value` on Scheduler, but for the sake of brevity:
        receiver.set_value(result);
    }
};

template <typename U, typename T, Sender<T> S>
struct map_sender {
    S prev_sender;
    function<U(T)> f;

    auto connect(Receiver<U> auto receiver) {
        return prev_sender.connect( map_receiver{ receiver, f } );
    }
};

// example usage:
on_thread_sender<int> sender1{ []{
    // think really hard
    return 42;
} };
auto sender2 = map_sender(sender1, [](int x) { return "answer=" + to_string(x); });
string result = get_blocking(sender2);
```

Other sequential transforms are implemented in pretty much [the same manner](https://github.com/UsatiyNyan/serious-execution-library/tree/main/include/sl/exec/algo/tf/seq). Also, important to point out that Sender composition does not increase synchronization cost, as opposed to previous implementation of the Future.

---

see also [algo_test](https://github.com/UsatiyNyan/serious-execution-library/blob/main/test/src/algo_test.cpp) for more examples
